demo/grid_table.py

```diff
@@ -0,0 +1,558 @@
+"""Pure Mesop Table built using CSS Grid.
+
+Functionality:
+
+- Column sorting
+- Header styles
+- Row styles
+- Cell styles
+- Cell templating
+- Row click
+- Expandable rows
+- Sticky header
+- Filtering (technically not built-in to the grid table component)
+
+TODOs:
+
+- Pagination
+- Sticky column
+- Control column width
+- Column filtering within grid table
+"""
+
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Any, Callable, Literal
+
+import pandas as pd
+
+import mesop as me
+
+df = pd.DataFrame(
+  data={
+    "NA": [pd.NA, pd.NA, pd.NA],
+    "Index": [3, 2, 1],
+    "Bools": [True, False, True],
+    "Ints": [101, 90, -55],
+    "Floats": [1002.3, 4.5, -1050203.021],
+    "Date Times": [
+      pd.Timestamp("20180310"),
+      pd.Timestamp("20230310"),
+      datetime(2023, 1, 1, 12, 12, 1),
+    ],
+    "Strings": ["Hello", "World", "!"],
+  }
+)
+
+SortDirection = Literal["asc", "desc"]
+
+
+@me.stateclass
+class State:
+  expanded_row_index: int | None = None
+  sort_column: str
+  sort_direction: SortDirection = "asc"
+  string_output: str
+  table_filter: str
+
+
+@me.page(
+  security_policy=me.SecurityPolicy(
+    allowed_iframe_parents=["https://google.github.io"]
+  ),
+  path="/grid_table",
+)
+def app():
+  state = me.state(State)
+
+  with me.box(style=me.Style(margin=me.Margin.all(30))):
+    # Simple example of filtering a data table. This is implemented separately of the
+    # grid table component. For simplicity, we only filter against a single column.
+    me.input(
+      label="Filter by Strings column",
+      style=me.Style(width="100%"),
+      on_blur=on_filter_by_strings,
+      on_enter=on_filter_by_strings,
+    )
+
+    # Grid Table demonstrating all features.
+    grid_table(
+      get_data_frame(),
+      header_config=GridTableHeader(sticky=True),
+      on_click=on_table_cell_click,
+      on_sort=on_table_sort,
+      row_config=GridTableRow(
+        columns={
+          "Bools": GridTableColumn(component=bool_component),
+          "Date Times": GridTableColumn(component=date_component),
+          "Floats": GridTableColumn(component=floats_component),
+          "Ints": GridTableColumn(style=ints_style, sortable=True),
+          "Strings": GridTableColumn(
+            component=strings_component, sortable=True
+          ),
+        },
+        expander=GridTableExpander(
+          component=expander,
+          row_index=state.expanded_row_index,
+        ),
+        style=striped_rows_style,
+      ),
+      sort_column=state.sort_column,
+      sort_direction=state.sort_direction,
+    )
+
+    # Used for demonstrating "table button" click example.
+    if state.string_output:
+      with me.box(
+        style=me.Style(
+          background="#141d2c",
```

Comment:
not the most important thing, but I think it would be nice to use "light theme" colors or have a toggle-able option between light and dark theme. (most people are going to copy this directly, and it's going to be kinda hard for them to adjust all the colors in a reasonably visually appealling way).

Reply:
That's a good point. I added a theme class for controlling the main theme. And added examples for light and dark.